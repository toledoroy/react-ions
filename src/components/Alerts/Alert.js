import React, { Component } from 'react'
import { bool, func, number, oneOf, string } from 'prop-types'
import Icon from '../Icon'
import StyledDiv from '../StyledDiv'
import { alertWrapper, closeIcon, countdownBar, countdownBarWrapper } from './styles.css'

class Alert extends Component {
  constructor(props) {
    super(props)
  }

  static propTypes = {
    /**
     * A class to add to the alert component.
     */
    className: string,
    /**
     * Whether the alert can be closed.
     */
    closable: bool,
    /**
     * Whether to show the countdown bar.
     */
    countdownBar: bool,
    /**
     * A callback to be triggered when the close icon is clicked or when the timeout expires.
     */
    onClose: func,
    /**
     * Whether or not to slide the alert in from the right.
     */
    slideIn: bool,
    /**
     * How long before the alert disappears.
     */
    timeout: number,
    /**
     * The alert type.
     */
    type: oneOf(['success', 'warning', 'info', 'danger'])
  }

  static defaultProps = {
    closable: true,
    countdownBar: true,
    type: 'success'
  }

  state = {
    timerStart: 0,
    timeout: this.props.timeout,
    timer: false
  }

  componentDidMount = () => {
    this.startTimer()
  }

  startTimer = () => {
    if (!this.props.timeout) return 

    this.setState({
      timerStart: new Date(),
      timer: setTimeout(this.closeAlert, this.state.timeout)
    })
  }

  pauseTimer = () => {
    if (!this.state.timer) return

    clearTimeout(this.state.timer)

    let timeout = this.state.timeout
    timeout -= new Date() - this.state.timerStart

    this.setState({ timeout })
  }

  closeAlert = () => {
    if (this.state.timer) clearTimeout(this.state.timer)
    if (this.props.onClose) this.props.onClose()
  }

  render = () => (
    <StyledDiv className={this.props.className} css={alertWrapper(this.props.type, this.props.closable, this.props.slideIn)} onMouseEnter={this.pauseTimer} onMouseLeave={this.startTimer}>
      <Icon name={`md-${this.props.type}`} width='17' height='17' />
      <div>{this.props.children}</div>
      {
        this.props.closable &&
        <StyledDiv css={closeIcon} onClick={this.closeAlert}>
          <Icon name='md-close' width='12' height='12' />
        </StyledDiv>
      }
      {
        this.props.countdownBar && !!this.props.timeout &&
        <StyledDiv css={countdownBarWrapper}>
          <StyledDiv className='countdown-bar' css={countdownBar(this.props.type, this.props.timeout)}></StyledDiv>
        </StyledDiv>
      }
    </StyledDiv>
  )
}

export default Alert
