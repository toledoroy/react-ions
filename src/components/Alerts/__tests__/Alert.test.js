import React from 'react'
import Alert from '../Alert'

describe('Alert', () => {
  describe('render', () => {
    it('should render a default alert (success)', () => {
      expect(shallow(<Alert timeout={5000}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should render a success alert', () => {
      expect(shallow(<Alert type='success' timeout={5000}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should render a warning alert', () => {
      expect(shallow(<Alert type='warning' timeout={5000}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should render an info alert', () => {
      expect(shallow(<Alert type='info' timeout={5000}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should render a danger alert', () => {
      expect(shallow(<Alert type='danger' timeout={5000}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should not render the countdown bar if countdownBar prop is set to false', () => {
      expect(shallow(<Alert type='success' timeout={5000} countdownBar={false}>Test alert</Alert>)).toMatchSnapshot()
    })
  
    it('should not render the countdown bar if timeout prop is not provided', () => {
      expect(shallow(<Alert>Test alert</Alert>)).toMatchSnapshot()
    })
  })

  describe('componentDidMount', () => {
    it('should start the timer when mounted', () => {
      const inst = shallow(<Alert>Test alert</Alert>).instance()
      const startTimer = sinon.spy(inst, 'startTimer')

      inst.componentDidMount()
      expect(startTimer.calledOnce).toBe.true
    })
  })

  describe('startTimer', () => {
    it('should not start the timer if the timeout prop is not provided', () => {
      const wrapper = shallow(<Alert>Test alert</Alert>)
      const state = wrapper.state()
      
      wrapper.instance().startTimer()
      expect(wrapper.state()).toEqual(state)
    })

    it('should start the timer if the timeout prop is provided', () => {
      const clock = sinon.useFakeTimers()
      const wrapper = shallow(<Alert timeout={5000}>Test alert</Alert>)
      const inst = wrapper.instance()
      const closeAlert = sinon.spy(inst, 'closeAlert')
      const state = wrapper.state()
      
      inst.startTimer()
      const updatedState = wrapper.state()
      expect(updatedState).not.toEqual(state)
      expect(updatedState).toMatchSnapshot()

      clock.tick(5000)
      expect(closeAlert.calledOnce).toBe.true

      clock.restore()
    })
  })

  describe('pauseTimer', () => {
    it('should not do anything if timer is not set in state', () => {
      const wrapper = shallow(<Alert>Test alert</Alert>)
      const state = wrapper.state()
      
      wrapper.instance().pauseTimer()
      expect(wrapper.state()).toEqual(state)
    })

    it('should pause the timer if timer is set in state', () => {
      const clock = sinon.useFakeTimers()
      const clearTimeout = sinon.spy(clock, 'clearTimeout')
      const wrapper = shallow(<Alert timeout={5000}>Test alert</Alert>)
      const inst = wrapper.instance()
      const state = wrapper.state()

      clock.tick(2000)
      
      inst.pauseTimer()
      const updatedState = wrapper.state()
      expect(updatedState).not.toEqual(state)
      expect(updatedState).toMatchSnapshot()

      expect(clearTimeout.calledOnce).toBe.true

      clock.restore()
    })
  })

  describe('closeAlert', () => {
    it('should not call clearTimeout if timer is not set in state', () => {
      const clock = sinon.useFakeTimers()
      const clearTimeout = sinon.spy(clock, 'clearTimeout')
      const inst = shallow(<Alert>Test alert</Alert>).instance()
      
      inst.closeAlert()

      expect(clearTimeout.called).toBe.false

      clock.restore()
    })

    it('should call clearTimeout if timer is set in state', () => {
      const clock = sinon.useFakeTimers()
      const clearTimeout = sinon.spy(clock, 'clearTimeout')
      const inst = shallow(<Alert timeout={5000}>Test alert</Alert>).instance()
      
      inst.closeAlert()

      expect(clearTimeout.calledOnce).toBe.true

      clock.restore()
    })

    it('should call the onClose prop if provided', () => {
      const clock = sinon.useFakeTimers()
      const clearTimeout = sinon.spy(clock, 'clearTimeout')
      const onClose = sinon.spy()
      const inst = shallow(<Alert timeout={5000} onClose={onClose}>Test alert</Alert>).instance()
      
      inst.closeAlert()

      expect(clearTimeout.calledOnce).toBe.true
      expect(onClose.calledOnce).toBe.true

      clock.restore()
    })
  })
})
