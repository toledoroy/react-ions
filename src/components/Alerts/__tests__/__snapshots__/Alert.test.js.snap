// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert pauseTimer should pause the timer if timer is set in state 1`] = `
Object {
  "timeout": 3000,
  "timer": 3,
  "timerStart": 1970-01-01T00:00:00.000Z,
}
`;

exports[`Alert render should not render the countdown bar if countdownBar prop is set to false 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#26B894",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-success"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should not render the countdown bar if timeout prop is not provided 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#26B894",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-success"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should render a danger alert 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#E54C3B",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-danger"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
  <.sc-bdVaJa
    css={
      Object {
        "bottom": 0,
        "height": "2px",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <.sc-bdVaJa
      className="countdown-bar"
      css={
        Array [
          "
  animation: ",
          "
  ",
          Keyframes {
            "id": "sc-keyframes-xYoGr",
            "inject": [Function],
            "name": "xYoGr",
            "rules": Array [
              "@-webkit-keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
              "@keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
            ],
            "toString": [Function],
          },
          " ",
          "5",
          "s linear 1;
",
          ";
  height: 100%;
  position: absolute;
  background-color: ",
          "#E54C3B",
          "
",
        ]
      }
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should render a default alert (success) 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#26B894",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-success"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
  <.sc-bdVaJa
    css={
      Object {
        "bottom": 0,
        "height": "2px",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <.sc-bdVaJa
      className="countdown-bar"
      css={
        Array [
          "
  animation: ",
          "
  ",
          Keyframes {
            "id": "sc-keyframes-xYoGr",
            "inject": [Function],
            "name": "xYoGr",
            "rules": Array [
              "@-webkit-keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
              "@keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
            ],
            "toString": [Function],
          },
          " ",
          "5",
          "s linear 1;
",
          ";
  height: 100%;
  position: absolute;
  background-color: ",
          "#26B894",
          "
",
        ]
      }
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should render a success alert 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#26B894",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-success"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
  <.sc-bdVaJa
    css={
      Object {
        "bottom": 0,
        "height": "2px",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <.sc-bdVaJa
      className="countdown-bar"
      css={
        Array [
          "
  animation: ",
          "
  ",
          Keyframes {
            "id": "sc-keyframes-xYoGr",
            "inject": [Function],
            "name": "xYoGr",
            "rules": Array [
              "@-webkit-keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
              "@keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
            ],
            "toString": [Function],
          },
          " ",
          "5",
          "s linear 1;
",
          ";
  height: 100%;
  position: absolute;
  background-color: ",
          "#26B894",
          "
",
        ]
      }
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should render a warning alert 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#F29C1E",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-warning"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
  <.sc-bdVaJa
    css={
      Object {
        "bottom": 0,
        "height": "2px",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <.sc-bdVaJa
      className="countdown-bar"
      css={
        Array [
          "
  animation: ",
          "
  ",
          Keyframes {
            "id": "sc-keyframes-xYoGr",
            "inject": [Function],
            "name": "xYoGr",
            "rules": Array [
              "@-webkit-keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
              "@keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
            ],
            "toString": [Function],
          },
          " ",
          "5",
          "s linear 1;
",
          ";
  height: 100%;
  position: absolute;
  background-color: ",
          "#F29C1E",
          "
",
        ]
      }
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert render should render an info alert 1`] = `
<.sc-bdVaJa
  css={
    Array [
      "
  display: flex;
  padding: 7px 10px;
  border-radius: 4px;
  background-color: ",
      "#fff",
      ";
  color: ",
      "#243140",
      ";
  font-size: 13px;
  line-height: 1.5;
  font-weight: 200;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.2), 0px 0px 1px 0px rgba(0, 0, 0, 0.2);
  margin-bottom: 15px;
  position: relative;
  overflow: hidden;

  ",
      "
  
  > svg {
    fill: ",
      "#62CAF3",
      ";
    margin-top: 1px;
    min-width: 17px;

    + div {
      margin-left: 10px;
      margin-right: ",
      "20px",
      ";
    }
  }

  &:hover .countdown-bar {
    animation-play-state: paused;
  }
",
    ]
  }
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <Icon
    height="17"
    name="md-info"
    width="17"
  />
  <div>
    Test alert
  </div>
  <.sc-bdVaJa
    css={
      Object {
        "> svg": Object {
          "fill": "#9198A0",
          "marginTop": 0,
        },
        "cursor": "pointer",
        "height": "15px",
        "margin": 0,
        "position": "absolute",
        "right": "10px",
        "top": "9px",
      }
    }
    onClick={[Function]}
  >
    <Icon
      height="12"
      name="md-close"
      width="12"
    />
  </.sc-bdVaJa>
  <.sc-bdVaJa
    css={
      Object {
        "bottom": 0,
        "height": "2px",
        "left": 0,
        "margin": 0,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <.sc-bdVaJa
      className="countdown-bar"
      css={
        Array [
          "
  animation: ",
          "
  ",
          Keyframes {
            "id": "sc-keyframes-xYoGr",
            "inject": [Function],
            "name": "xYoGr",
            "rules": Array [
              "@-webkit-keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
              "@keyframes xYoGr{0%{width:100%;}100%{width:0%;}}",
            ],
            "toString": [Function],
          },
          " ",
          "5",
          "s linear 1;
",
          ";
  height: 100%;
  position: absolute;
  background-color: ",
          "#62CAF3",
          "
",
        ]
      }
    />
  </.sc-bdVaJa>
</.sc-bdVaJa>
`;

exports[`Alert startTimer should start the timer if the timeout prop is provided 1`] = `
Object {
  "timeout": 5000,
  "timer": 2,
  "timerStart": 1970-01-01T00:00:00.000Z,
}
`;
